### PR TITLE
[MI-3561] Fixed issue: Getting multiple DMs when Oauth token is revoked

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**
- Created a single API to get the data for LHS and todo command.

Issue Link: https://github.com/mattermost/mattermost-plugin-gitlab/issues/371

### CODE REVIEW

File: server/api.go [hunk 1]
```diff
@@ -615,10 +551,10 @@ func (p *Plugin) getPrDetails(c *UserContext, w http.ResponseWriter, r *http.Req
 	p.writeAPIResponse(w, result)
 }
 
-func (p *Plugin) getYourAssignments(c *UserContext, w http.ResponseWriter, r *http.Request) {
-	var result []*gitlab.Issue
+func (p *Plugin) getLHSData(c *UserContext, w http.ResponseWriter, r *http.Request) {
```
[the remark this answers is not in the document]
Comment (raghavaggarwal2308) [reply server/api.go:554]: Discussed with @Kshitij-Katiyar

**Review by ayusht2810 — APPROVED** [review]

File: server/gitlab/api.go [hunk 2]
```diff
@@ -282,14 +282,61 @@ func (g *gitlab) GetProject(ctx context.Context, user *UserInfo, token *oauth2.T
 	return project, nil
 }
 
-func (g *gitlab) GetReviews(ctx context.Context, user *UserInfo, token *oauth2.Token) ([]*MergeRequest, error) {
+type SidebarContent struct {
```
Comment (Kshitij-Katiyar) [inline server/gitlab/api.go:285]:
```suggestion
type LHSContent struct {
```

File: server/gitlab/api.go [hunk 3]
```diff
@@ -518,14 +561,10 @@ func (g *gitlab) fetchYourPrDetails(c context.Context, log logger.Logger, client
...
-func (g *gitlab) GetYourAssignments(ctx context.Context, user *UserInfo, token *oauth2.Token) ([]*Issue, error) {
-	client, err := g.GitlabConnect(*token)
-	if err != nil {
-		return nil, err
-	}
+func (g *gitlab) GetYourAssignments(ctx context.Context, user *UserInfo, client *internGitlab.Client) ([]*Issue, error) {
 	opened := stateOpened
 	scope := scopeAll
-
+	var err error
```
Comment (Kshitij-Katiyar) [inline server/gitlab/api.go:567]: why we have added this ?

File: server/plugin.go [hunk 4]
```diff
@@ -708,10 +651,13 @@ func (p *Plugin) GetToDo(ctx context.Context, user *gitlab.UserInfo) (bool, stri
 			hasTodo = true
 		}
 
+		if err := g.Wait(); err != nil {
```
Comment (Kshitij-Katiyar) [inline server/plugin.go:654]: Why we have added this here ?

File: webapp/src/components/sidebar_buttons/index.js [hunk 5]
```diff
@@ -20,10 +17,10 @@ function mapStateToProps(state) {
         connected: state[`plugins-${id}`].connected,
         username: state[`plugins-${id}`].username,
         clientId: state[`plugins-${id}`].clientId,
-        reviews: state[`plugins-${id}`].reviews,
-        yourPrs: state[`plugins-${id}`].yourPrs,
-        yourAssignments: state[`plugins-${id}`].yourAssignments,
-        unreads: state[`plugins-${id}`].unreads,
+        reviews: state[`plugins-${id}`].lhsData.reviews,
```
Comment (Kshitij-Katiyar) [inline webapp/src/components/sidebar_buttons/index.js:20]: optional changing required here ?

File: server/gitlab/api.go [hunk 6]
```diff
@@ -322,7 +368,7 @@ func (g *gitlab) GetReviews(ctx context.Context, user *UserInfo, token *oauth2.T
 		for {
 			var current []*internGitlab.MergeRequest
 			var resp *internGitlab.Response
-			current, resp, err = client.MergeRequests.ListGroupMergeRequests(g.gitlabGroup, opt)
+			current, resp, err := client.MergeRequests.ListGroupMergeRequests(g.gitlabGroup, opt)
```
Comment (Kshitij-Katiyar) [inline server/gitlab/api.go:371]: There are 2 declarations of the resp.

File: server/gitlab/api.go [hunk 7]
```diff
@@ -338,7 +384,7 @@ func (g *gitlab) GetReviews(ctx context.Context, user *UserInfo, token *oauth2.T
 	for _, mr := range mrs {
 		if mr.Labels != nil {
 			var labelsWithDetails []*internGitlab.Label
-			labelsWithDetails, err = g.GetLabelDetails(client, mr.ProjectID, mr.Labels)
+			labelsWithDetails, err := g.GetLabelDetails(client, mr.ProjectID, mr.Labels)
```
Comment (Kshitij-Katiyar) [inline server/gitlab/api.go:387]: same

File: server/gitlab/api.go [hunk 8]
```diff
@@ -574,7 +608,7 @@ func (g *gitlab) GetYourAssignments(ctx context.Context, user *UserInfo, token *
 	for _, issue := range issues {
 		if issue.Labels != nil {
 			var labelsWithDetails []*internGitlab.Label
-			labelsWithDetails, err = g.GetLabelDetails(client, issue.ProjectID, issue.Labels)
+			labelsWithDetails, err := g.GetLabelDetails(client, issue.ProjectID, issue.Labels)
```
Comment (Kshitij-Katiyar) [inline server/gitlab/api.go:611]: same

File: server/gitlab/api.go [hunk 9]
```diff
@@ -588,20 +622,16 @@ func (g *gitlab) GetYourAssignments(ctx context.Context, user *UserInfo, token *
...
-
+func (g *gitlab) GetUnreads(ctx context.Context, user *UserInfo, client *internGitlab.Client) ([]*internGitlab.Todo, error) {
 	var todos []*internGitlab.Todo
+
 	opt := &internGitlab.ListTodosOptions{
 		ListOptions: internGitlab.ListOptions{Page: 1, PerPage: perPage},
 	}
 	for {
 		var current []*internGitlab.Todo
 		var resp *internGitlab.Response
```
Comment (Kshitij-Katiyar) [inline server/gitlab/api.go:633]: same

File: server/gitlab/api.go [hunk 10]
```diff
@@ -282,12 +282,58 @@ func (g *gitlab) GetProject(ctx context.Context, user *UserInfo, token *oauth2.T
 	return project, nil
 }
 
-func (g *gitlab) GetReviews(ctx context.Context, user *UserInfo, token *oauth2.Token) ([]*MergeRequest, error) {
+type LHSContent struct {
```
Comment (Kshitij-Katiyar) [inline server/gitlab/api.go:285]: See this plugin have serializers package and initialise this there. If not then the top of this file.

**Review by Kshitij-Katiyar — APPROVED** [review]

**Pull request merged**